odl/tomo/geometry/conebeam.py

```diff
@@ -1195,24 +1309,57 @@ def src_position(self, angle):
         True
         >>> geom.src_position(np.zeros((4, 5))).shape
         (4, 5, 3)
+
+        Specifying flying focal spot:
+
+        >>> apart = odl.uniform_partition(0, 2 * np.pi, 4)
+        >>> geom = ConeBeamGeometry(
+        ...     apart, dpart,
+        ...     src_radius=1, det_radius=5,
+        ...     src_shift_func=lambda angle: flying_focal_spot(
+        ...         angle,
+        ...         apart=apart,
+        ...         shifts=[(0, 0.1, 0), (0, 0, 0.1)]),
+        ...         src_to_det_init=(-0.71, 0.71, 0))
+        >>> geom.angles
+        array([ 0.78539816,  2.35619449,  3.92699082,  5.49778714])
+        >>> np.round(geom.src_position(geom.angles), 2)
+        array([[ 1. ,  0.1,  0. ],
+               [ 0. ,  1. ,  0.1],
+               [-1. , -0.1,  0. ],
+               [-0. , -1. ,  0.1]])
         """
         squeeze_out = (np.shape(angle) == ())
         angle = np.array(angle, dtype=float, copy=False, ndmin=1)
         rot_matrix = self.rotation_matrix(angle)
         extra_dims = angle.ndim
+        src_shifts = self.src_shift_func(angle)
 
         # Initial vector from center of rotation to source.
         # It can be computed this way since source and detector are at
         # maximum distance, i.e. the connecting line passes the center.
         center_to_src_init = -self.src_radius * self.src_to_det_init
+        # shifting the source according to ffs
+        tangent = -np.cross(-self.src_to_det_init, self.axis)
+        tangent /= np.linalg.norm(tangent)
+        ffs_shift = (np.multiply.outer(src_shifts[:, 0],
+                                       -self.src_to_det_init)
+                     + np.multiply.outer(src_shifts[:, 1], tangent))
+        center_to_src_init = center_to_src_init + ffs_shift
+        # broadcasting to perform matrix multiplication "manually",
+        # since existing numpy functions do cross product along the outer
+        # dimensions, which we don't need here
+        center_to_src_init = np.repeat(center_to_src_init, 3,
+                                       axis=-2).reshape(-1, 3, 3)
         # `circle_component` has shape (a, ndim)
-        circle_component = rot_matrix.dot(center_to_src_init)
+        circle_component = np.sum(rot_matrix * center_to_src_init, axis=-1)
```

Comment:
Can't we do that with `np.matmul` somehow? It behaves differently from `np.dot`.

Reply:
I don't think so, since the arrays are 3 dimensional here

rot_matrix has dimension (n_angles, 3, 3)
center_to_src_init initially has dimension (n_angles, 3)

For each angles we need to multiply a vector by a matrix, but we don't need to "cross- multiply" vectors and matrices corresponding to different angles. Documentation for matmul says:
"Stacks of matrices are broadcast together as if the matrices were elements, respecting the signature (n,k),(k,m)->(n,m):"

So, it does not handle 3d arrays the way we need it here.

Reply:
I agree that the way I did it looks clumsy, it took me quite some time to understand what I did there and why =D

Reply:
Haha, no worries. If it works and there's no obvious easier way it's fine. I was just wondering.

Reply:
Use a einsum. I think this should be

```
np.einsum('bji, bi-> bj', rot_matrix, center_to_src_init)
```